### DISCUSSION TITLE
💭python輯運算的解讀

### DISCUSSION THREAD

**Discussion by JunTingLin** [1101-GS4719-C/Python-Community-GS4719-C#463] [General]
## 前言
看到這位大大po了一篇邏輯運算的解讀[寫code真的要注意你寫哪個語言! #449](https://github.com/NCU-GS4719-Python/Python-Community/discussions/449)

原po在裡面寫了 ⏬
```
a=2
b="Apple"
c=""
print(a or 0) #會印2
print(b and True)#會印True(因為邏輯判斷原則)
print(True and b)#會印Apple
print(c and True)#會印""
```
可能很多人都知道原理了，但我那時還是沒有很了解😅，b and True，為何出來的是True，可是True and b出來是b的"Apple"🤔

自己試了幾次，並且查詢官方文件[Boolean operations](https://docs.python.org/3/reference/expressions.html#boolean-operations)
整理並與大家分享😀

## 是False的值有哪些?
False值包括False、None和0

## not 運算子
```
print(not 0) #not False #會印True
print(not 2) #not True #會印False
print(not None) #not False #會印True
```

## or看到True就會短路
or左右兩邊均為True，回傳，Python 會立即返回那個“真”值
```
print(2 or 9) #True or True #會印2
print(9 or 2) #True or True #會印9
```

```
print(0 or 9) #False or True #會印9
print(9 or 0) #True or False #會印9
```

## or沒遇到True，看最後
or如果它完成並且沒有任何內容為真，它會返回它處理的最後一個參數，這將是一個“假”值
```
print(0 or None) #False and False #會印None
```

## and看到False就會短路
and如果它看到任何假值，它會立即以那個“假”值退出
```
print(0 and None) #False and False #會印0
print(None and 0) #False and False #會印None
```

```
print(2 and None) #True and False #會印None
print(None and 2) #False and True #會印None
```

## and沒遇到False，看最後
and如果它完成並且沒有任何內容為假值，它會返回它處理的最後一個參數，這將是一個“真”值
```
print(2 and 3) #True and True #會印3
```

## 整理🔆
+ or
看到True我就印它(因為短路了後面也沒有必要看了)
假如都沒有True(兩個都是False)，我印最後一個

+ and
看到False我就印它(因為短路了)
假如都沒有False(兩個都是True)，我印最後一個

# 小節
短路現象當然並不是只有python專有，java、C++都有，但是`2 and 9` 這種寫法應該只有python、Ruby特有💦

# 參考
+ [Stackoverflow - False or None vs. None or False](https://stackoverflow.com/questions/3914667/false-or-none-vs-none-or-false)

+ [docs - Boolean operations](https://docs.python.org/3/reference/expressions.html#boolean-operations)

**Comment by ghost** [comment]
原來如此0.0有稍微理解整個邏輯了><

**Comment by ThomasMeng0515** [comment]
之前第一次碰到也覺得滿奇怪的，內文所提到的"短路"應該就是老師上課說的"快速運算"吧~
這樣可以節省電腦的資源。
你的整理讓我更了解了!

**Comment by chupochun** [comment]
謝謝分享，讓我複習了上一周的內容

**Comment by sean1325418** [comment]
只要記得and會找False，or會找True，其他時候都是印最後就好

**Comment by xkllkx** [comment]
謝謝分享，幫我複習了邏輯判斷的部分。
